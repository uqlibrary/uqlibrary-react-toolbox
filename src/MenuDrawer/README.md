# MenuDrawer

simple reusable menu for UQ Library applications using ReactJs

## Props
- `menuItems: PropTypes.array.isRequired` - a list of menu items, see usage
- `drawerOpen: PropTypes.bool`
- `docked: PropTypes.bool`
- `toggleDrawer: PropTypes.func`
- `logoImage: PropTypes.string` - url of an image of a logo
- `logoText: PropTypes.string` - alt text for a logo image
    
## Usage

**./App.js**
```jsx
import React from 'react';
...

import {MenuDrawer} from 'uqlibrary-react-toolbox';

// Icon import is required to display an icon in the menu
import ActionHome from 'material-ui/svg-icons/action/home';


// Menu configuration
const menuItems = [
            {
                linkTo: '/',
                primaryText: 'Home',
                leftIcon: <ActionHome />,
            },
            {
                divider: true
            },
            {
                linkTo: '/Help',
                primaryText: 'Help'
            }
        ];
        
const App = () => (
  <div>
    <MenuDrawer 
      menuItems={menuItems}
      logoImage='http://images/image.svg'
      logoText='logo image'
      drawerOpen={docked ? true : drawerOpen}
      docked={docked}
      toggleDrawer={this.toggleDrawer}
      imgAltText: 'Image Alt Text',
      imgLogo: 'https://url/to/image'
       />
  </div>
);

ReactDOM.render(
  <App />,
  document.getElementById('app')
);
```

