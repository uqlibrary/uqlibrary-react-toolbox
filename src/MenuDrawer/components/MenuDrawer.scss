@import '../sass/variables';

.main-drawer {
  .side-drawer {
    .logo-wrapper {
      height: 75px;
      padding: 10px;
      text-align: center;
      background-color: $uq-primary-color;
      -webkit-box-sizing: border-box;
      -moz-box-sizing: border-box;
      box-sizing: border-box;
      width: 100%;

      img {
        height: 100%;
      }
    }

    .main-menu {
      padding-top: 0 !important;

      a {
        text-decoration: none;
      }
    }
  }
}
