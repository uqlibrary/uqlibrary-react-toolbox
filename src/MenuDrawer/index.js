export {default as MenuDrawer} from './MenuDrawer';
