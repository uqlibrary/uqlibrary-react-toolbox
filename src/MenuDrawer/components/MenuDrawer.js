import React from 'react';
import { Link } from 'react-router-dom';
import {List, ListItem} from 'material-ui/List';
import Divider from 'material-ui/Divider';
import Drawer from 'material-ui/Drawer';

import './MenuDrawer.scss';

const propTypes = {
    menuItems: React.PropTypes.array.isRequired,
    drawerOpen: React.PropTypes.bool,
    docked: React.PropTypes.bool,
    toggleDrawer: React.PropTypes.func
};

const uqLogo = 'https://static.uq.net.au/v1/logos/corporate/uq-logo-white.svg';
const MenuDrawer = ({menuItems, toggleDrawer, drawerOpen, docked}) => {
    return (
        <Drawer
            containerClassName="main-drawer flex"
            open={drawerOpen}
            width={320}
            onRequestChange={() => toggleDrawer(!drawerOpen)}
            docked={docked}>
            <div className="layout-fill side-drawer">
                <div className="logo-wrapper">
                    <img src={uqLogo} alt="University of Queensland logo"/>
                </div>
                <List className="main-menu">
                    {menuItems.map((menuItem, index) =>
                    <span className="menu-item-container" key={index}>
                        {menuItem.divider ? (
                            <Divider />
                            ) : (
                            <Link to={menuItem.linkTo}>
                                <ListItem primaryText={menuItem.primaryText}
                                          secondaryText={menuItem.secondaryText}
                                          leftIcon={menuItem.leftIcon ? menuItem.leftIcon : null}
                                />
                            </Link>
                        )}
                    </span>
                    )}
                </List>
            </div>
        </Drawer>
    );
};

MenuDrawer.propTypes = propTypes;

export default MenuDrawer;
