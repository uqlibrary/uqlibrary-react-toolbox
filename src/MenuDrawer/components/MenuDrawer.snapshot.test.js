jest.dontMock('./MenuDrawer');

import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import MenuDrawer from './MenuDrawer';

function setup(drawerOpen, menuItems, docked) {

    const defaultMenuItems = [
        {
            linkTo: '/',
            primaryText: 'Primary text 0',
            secondaryText: 'secondary text 0'
        },
        {
            divider: true
        },
        {
            linkTo: '/xyz',
            primaryText: 'Primary text 1',
            secondaryText: 'secondary text 1'
        }
    ];


    const props = {
        menuItems: menuItems ? menuItems : defaultMenuItems,
        drawerOpen: drawerOpen,
        docked: docked,
        toggleDrawer: () => { },
    };

    return shallow(<MenuDrawer {...props} />);
}

describe('MenuDrawer snapshots tests', () => {
    it('renders menu with dividers', () => {
        const wrapper = setup();
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });
    it('renders menu without dividers', () => {
        const menuItems = [
            {
                linkTo: '/',
                primaryText: 'Primary text 0',
                secondaryText: 'secondary text 0'
            },
            {
                linkTo: '/abc',
                primaryText: 'Primary text 0',
                secondaryText: 'secondary text 0'
            }
        ];
        const wrapper = setup(false, menuItems, false);
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });
});
