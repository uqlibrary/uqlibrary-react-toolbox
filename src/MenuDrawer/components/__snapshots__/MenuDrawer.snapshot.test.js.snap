// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuDrawer snapshots tests renders menu with dividers 1`] = `
<Drawer
  containerClassName="main-drawer"
  disableSwipeToOpen={false}
  docked={true}
  onRequestChange={[Function]}
  open={null}
  openSecondary={false}
  swipeAreaWidth={30}
  width={320}
  zDepth={2}
>
  <div
    className="layout-fill side-drawer"
  >
    <div
      className="logo-wrapper"
    >
      <div
        className="columns is-gapless is-mobile"
      >
        <div
          className="column is-centered"
        />
        <div
          className="column is-narrow is-hidden-tablet menuCloseButton"
        >
          <IconButton
            disableTouchRipple={false}
            disabled={false}
            iconStyle={Object {}}
            onTouchTap={[Function]}
            tooltipPosition="bottom-center"
            touch={false}
          >
            <HardwareKeyboardArrowLeft />
          </IconButton>
        </div>
      </div>
    </div>
    <List
      className="main-menu"
      id="mainMenu"
      tabIndex={-1}
    >
      <div
        className="skipNav"
        onClick={[Function]}
        onKeyPress={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="skipButton"
        />
      </div>
      <span
        className="menu-item-container"
        key="0"
      >
        <ListItem
          autoGenerateNestedIndicator={true}
          containerElement="span"
          disableKeyboardFocus={false}
          disabled={false}
          initiallyOpen={false}
          insetChildren={false}
          leftIcon={null}
          nestedItems={Array []}
          nestedLevel={0}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onNestedListToggle={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          open={null}
          primaryText="Primary text 0"
          primaryTogglesNestedList={false}
          secondaryText="secondary text 0"
          secondaryTextLines={1}
          tabIndex={-1}
        />
      </span>
      <span
        className="menu-item-container"
        key="2"
      >
        <ListItem
          autoGenerateNestedIndicator={true}
          containerElement="span"
          disableKeyboardFocus={false}
          disabled={false}
          initiallyOpen={false}
          insetChildren={false}
          leftIcon={null}
          nestedItems={Array []}
          nestedLevel={0}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onNestedListToggle={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          open={null}
          primaryText="Primary text 1"
          primaryTogglesNestedList={false}
          secondaryText="secondary text 1"
          secondaryTextLines={1}
          tabIndex={-1}
        />
      </span>
    </List>
  </div>
</Drawer>
`;

exports[`MenuDrawer snapshots tests renders menu with logo 1`] = `
<Drawer
  containerClassName="main-drawer"
  disableSwipeToOpen={false}
  docked={false}
  onRequestChange={[Function]}
  open={false}
  openSecondary={false}
  swipeAreaWidth={30}
  width={320}
  zDepth={2}
>
  <div
    className="layout-fill side-drawer"
  >
    <div
      className="logo-wrapper"
    >
      <div
        className="columns is-gapless is-mobile"
      >
        <div
          className="column is-centered"
        >
          <img
            alt="desc of image"
            src="http://image/image.svg"
          />
        </div>
        <div
          className="column is-narrow is-hidden-tablet menuCloseButton"
        >
          <IconButton
            disableTouchRipple={false}
            disabled={false}
            iconStyle={Object {}}
            onTouchTap={[Function]}
            tooltipPosition="bottom-center"
            touch={false}
          >
            <HardwareKeyboardArrowLeft />
          </IconButton>
        </div>
      </div>
    </div>
    <List
      className="main-menu"
      id="mainMenu"
      tabIndex={-1}
    >
      <div
        className="skipNav"
        onClick={[Function]}
        onKeyPress={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="skipButton"
        />
      </div>
      <span
        className="menu-item-container"
        key="0"
      >
        <ListItem
          autoGenerateNestedIndicator={true}
          containerElement="span"
          disableKeyboardFocus={false}
          disabled={false}
          initiallyOpen={false}
          insetChildren={false}
          leftIcon={null}
          nestedItems={Array []}
          nestedLevel={0}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onNestedListToggle={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          open={null}
          primaryText="Primary text 0"
          primaryTogglesNestedList={false}
          secondaryText="secondary text 0"
          secondaryTextLines={1}
          tabIndex={-1}
        />
      </span>
      <span
        className="menu-item-container"
        key="1"
      >
        <ListItem
          autoGenerateNestedIndicator={true}
          containerElement="span"
          disableKeyboardFocus={false}
          disabled={false}
          initiallyOpen={false}
          insetChildren={false}
          leftIcon={null}
          nestedItems={Array []}
          nestedLevel={0}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onNestedListToggle={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          open={null}
          primaryText="Primary text 0"
          primaryTogglesNestedList={false}
          secondaryText="secondary text 0"
          secondaryTextLines={1}
          tabIndex={-1}
        />
      </span>
    </List>
  </div>
</Drawer>
`;

exports[`MenuDrawer snapshots tests renders menu without dividers 1`] = `
<Drawer
  containerClassName="main-drawer"
  disableSwipeToOpen={false}
  docked={false}
  onRequestChange={[Function]}
  open={false}
  openSecondary={false}
  swipeAreaWidth={30}
  width={320}
  zDepth={2}
>
  <div
    className="layout-fill side-drawer"
  >
    <div
      className="logo-wrapper"
    >
      <div
        className="columns is-gapless is-mobile"
      >
        <div
          className="column is-centered"
        />
        <div
          className="column is-narrow is-hidden-tablet menuCloseButton"
        >
          <IconButton
            disableTouchRipple={false}
            disabled={false}
            iconStyle={Object {}}
            onTouchTap={[Function]}
            tooltipPosition="bottom-center"
            touch={false}
          >
            <HardwareKeyboardArrowLeft />
          </IconButton>
        </div>
      </div>
    </div>
    <List
      className="main-menu"
      id="mainMenu"
      tabIndex={-1}
    >
      <div
        className="skipNav"
        onClick={[Function]}
        onKeyPress={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="skipButton"
        />
      </div>
      <span
        className="menu-item-container"
        key="0"
      >
        <ListItem
          autoGenerateNestedIndicator={true}
          containerElement="span"
          disableKeyboardFocus={false}
          disabled={false}
          initiallyOpen={false}
          insetChildren={false}
          leftIcon={null}
          nestedItems={Array []}
          nestedLevel={0}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onNestedListToggle={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          open={null}
          primaryText="Primary text 0"
          primaryTogglesNestedList={false}
          secondaryText="secondary text 0"
          secondaryTextLines={1}
          tabIndex={-1}
        />
      </span>
      <span
        className="menu-item-container"
        key="1"
      >
        <ListItem
          autoGenerateNestedIndicator={true}
          containerElement="span"
          disableKeyboardFocus={false}
          disabled={false}
          initiallyOpen={false}
          insetChildren={false}
          leftIcon={null}
          nestedItems={Array []}
          nestedLevel={0}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onNestedListToggle={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          open={null}
          primaryText="Primary text 0"
          primaryTogglesNestedList={false}
          secondaryText="secondary text 0"
          secondaryTextLines={1}
          tabIndex={-1}
        />
      </span>
    </List>
  </div>
</Drawer>
`;
